### PR TITLE
Fix recompilation of regexps during backtracking in the parser

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
If a query with a syntax error contained COLUMNS matcher with a regular expression, the regular expression was compiled each time during the parser's backtracking, instead of being compiled once. This was a fundamental error. The compiled regexp was put to AST. But the letter A in AST means "abstract" which means it should not contain heavyweight objects. Parts of AST can be created and discarded during parsing, including a large number of backtracking. This leads to slowness on the parsing side and consequently allows DoS by a readonly user. But the main problem is that it prevents progress in fuzzers.

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 613c31038ae76be09d35a151e6446024b07f127b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61543/613c31038ae76be09d35a151e6446024b07f127b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/03013_parser_regexp_recompilation.txt
Comment (serxa) [inline tests/queries/0_stateless/03013_parser_regexp_recompilation.txt:1]: it feels like a binary, not .txt file

**Review by serxa — APPROVED** [review]

File: tests/queries/0_stateless/03013_parser_regexp_recompilation.txt
Comment (alexey-milovidov) [reply tests/queries/0_stateless/03013_parser_regexp_recompilation.txt:1]: Yes, although the emphasis of naming it .txt is that it contains the query text (which has some binary characters). I think it is ok.

**Comment by alexey-milovidov** [comment]
I cannot explain the difference that the Performance test has shown and assume they are irrelevant.

**Pull request merged**